File: Week 10/Java/reverseVowels.java
/**
 * Prompt:
 * 
 * Write a function that takes a string as input and reverse only the vowels of
 * a string.
 * 
 * https://leetcode.com/problems/reverse-vowels-of-a-string/
 */

class Solution {
    public String reverseVowels(String s) {
        StringBuilder reversedVowels = new StringBuilder(s);

        int head = 0, tail = s.length() - 1;
        while (head < tail) {
            char headChar = reversedVowels.charAt(head);
            char tailChar = reversedVowels.charAt(tail);

            if (isVowel(headChar) && !isVowel(tailChar)) {
                tail--;
                continue;
            } else if (!isVowel(headChar) && isVowel(tailChar)) {
                head++;
                continue;
            } else if (isVowel(headChar) && isVowel(tailChar)) {
                reversedVowels.replace(head, head + 1, "" + tailChar);
                reversedVowels.replace(tail, tail + 1, "" + headChar);
            }

            head++;
            tail--;
        }

        return reversedVowels.toString();
    }

    boolean isVowel(char c) {
        return c == 'a' || c == 'e' || c == 'i' || c == 'o' || c == 'u' || c == 'A' || c == 'E' || c == 'I' || c == 'O'
                || c == 'U';
    }
}

/**
 * Not the best solution, runtime of 59ms (faster than 9.26%) and memory usage
 * of 43.5mb (less than 5.16%)
 */

File: First Job Hunt/Week 10/JavaScript/reverseVowels.js
/**
 * Prompt:
 *
 * Write a function that takes a string as input and reverse only the vowels of
 * a string.
 *
 * https://leetcode.com/problems/reverse-vowels-of-a-string/
 */

const reverseVowels = (s) => {
  let letters = s.split("");
  let i = 0;
  let j = letters.length - 1;
  let vowels = ["a", "e", "i", "o", "u", "A", "E", "I", "O", "U"];

  while (i < j) {
    let temp = "";
    if (vowels.includes(letters[i]) && vowels.includes(letters[j])) {
      temp = letters[i];
      letters[i] = letters[j];
      letters[j] = temp;
      i++;
      j--;
    } else if (!vowels.includes(letters[i])) {
      i++;
    } else {
      j--;
    }
  }

  return letters.join("");
};

/**
 * Fairly similar to my Java solution with much better stats. Runtime of 152ms (faster than 21.33%)
 * and memory usage of 43.9mb (less than 85.83%)
 */
